### PR TITLE
Refactor to make ACL errors more structured.

### PR DESCRIPTION
**Pull request by markan**
Port of the structured error work in consul-enterprise.

This is the OSS side of #12239

### CODE REVIEW

File: acl/errors.go [hunk 1]
```diff
@@ -61,18 +61,67 @@ func IsErrPermissionDenied(err error) bool {
 	return err != nil && strings.Contains(err.Error(), errPermissionDenied)
 }
 
+// Arguably this should be some sort of union type.
+// The usage of Cause and the rest of the fields is entirely disjoint.
```
Comment (dnephin) [inline acl/errors.go:65]: I'm pretty sure we'll be able to make this not entirely disjoint in the future! There are at least 2 places in enterprise where we'll want to use `Cause` along with the other fields. The current implementation doesn't support that yet, but I think that's ok. We can add support when it comes up.

File: acl/errors.go [hunk 2]
```diff
@@ -61,18 +61,67 @@ func IsErrPermissionDenied(err error) bool {
...
+// There are generally three classes of errors
+// 1) Named entities without a context
+// 2) Unnamed entities with a context
+// 3) Completely context free checks (global permissions)
+// 4) Errors that only have a cause and bad
 func (e PermissionDeniedError) Error() string {
+	var message strings.Builder
+	message.WriteString(errPermissionDenied)
+
+	// This is used where we
```
Comment (dnephin) [inline acl/errors.go:90]: Missing end of comment?

File: acl/errors.go [hunk 3]
```diff
@@ -61,18 +61,67 @@ func IsErrPermissionDenied(err error) bool {
...
+	// e.g. "sidecar-proxy-1"
+	ResourceID ResourceDescriptor
 }
 
+// Initially we may not have attribution information; that will become more complete as we work this change through
+// There are generally three classes of errors
+// 1) Named entities without a context
+// 2) Unnamed entities with a context
+// 3) Completely context free checks (global permissions)
+// 4) Errors that only have a cause and bad
```
Comment (dnephin) [inline acl/errors.go:85]: maybe missing a word on this last line? "and bad?"

**Review by dnephin — COMMENTED** [review]
Looks good! Just noticing a few minor things in comments

**Review by dnephin — APPROVED** [review]
LGTM!

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/581580.

**Pull request merged**